src/saveload/marker.rs

```diff
@@ -0,0 +1,213 @@
+///     // Entities with too old sequence id get deleted.
+///     fn update(&mut self, update: Self) {
+///         assert_eq!(self.id, update.id);
+///         self.seq = update.seq;
+///     }
+/// }
+///
+/// // Each client and server has one
+/// // Contains id range and `NetMarker -> Entity` mapping
+/// struct NetNode {
+///     range: Range<u64>,
+///     mapping: HashMap<u64, Entity>,
+/// }
+///
+/// impl MarkerAllocator<NetMarker> for NetNode {
+///     fn allocate(&mut self, entity: Entity, id: Option<u64>) -> NetMarker {
+///         let id = id.unwrap_or_else(|| {
+///             self.range.next().expect("Id range must be virtually endless")
+///         });
+///         let marker = NetMarker {
+///             id: id,
+///             seq: 0,
+///         };
+///         self.mapping.insert(id, entity);
+///         marker
+///     }
+///
+///     fn get(&self, id: u64) -> Option<Entity> {
+///         self.mapping.get(&id).cloned()
+///     }
+/// }
+///
+/// fn main() {}
+/// ```
+pub trait Marker: Component + DeserializeOwned + Serialize + Copy {
+    /// Id of the marker
+    type Identifier: Copy + Debug + Eq + Hash;
+
+    /// Allocator for this `Marker`
+    type Allocator: MarkerAllocator<Self>;
+
+    /// Get this marker internal id
+    fn id(&self) -> Self::Identifier;
+
+    /// Update marker with new value.
+    /// It must preserve internal `Identifier`.
+    ///
+    /// ## Panics
+    ///
+    /// Allowed to panic if `self.id() != update.id()`.
+    /// But usually implementer may ignore `update.id()` value
+    /// as deserialization algorithm ensures `id()`s match.
+    fn update(&mut self, update: Self) {
+        ::std::mem::drop(update);
+    }
+}
+
+/// This allocator is used with `Marker` trait.
+/// It provides method for allocation of `Marker`s.
+/// And also should provide `Marker -> Entity` mapping
+/// `maintain` method can be implemented for cleanup and actualization.
+/// See docs for `Marker` for example.
+pub trait MarkerAllocator<M: Marker>: Resource {
+    /// Allocate new `Marker`.
+    /// Stores mapping `Marker` -> `Entity`.
+    /// If _id_ argument is `Some(id)` then new marker will have this `id`.
+    /// Otherwise allocator creates marker with new unique id.
+    fn allocate(&mut self, entity: Entity, id: Option<M::Identifier>) -> M;
+
+    /// Get `Entity` by `Marker::Identifier`
+    fn get(&self, id: M::Identifier) -> Option<Entity>;
+
+    /// Create new unique marker `M` and attach it to entity.
+    /// Or get old marker if already marked.
+    fn mark<'a>(&mut self, entity: Entity, storage: &mut WriteStorage<'a, M>) -> M {
+        match storage.get(entity).cloned() {
+            Some(marker) => marker,
+            None => {
+                let marker = self.allocate(entity, None);
+                storage.insert(entity, marker);
+                marker
+            }
+        }
+    }
+
+    /// Find `Entity` by `Marker` with same id and update `Marker` attached instance.
```

Comment:
s/\`Entity\` by/an \`Entity\` by a/
s/\`Marker\` attached/the \`Marker\` attached to the/